gatewayconfig/gpio/mraa_button.py

```diff
@@ -0,0 +1,101 @@
+import time
+import threading
+
+from hm_pyhelper.logger import get_logger
+from gatewayconfig.gpio.mraa_gpio import mraa_gpio_write
+
+
+LOGGER = get_logger(__name__)
+
+try:
+    import mraa
+except:
+    LOGGER.warn("mraa not loaded. If you see this message outside of unit tests, this may cause problems.")
+    
+# How often to check for button presses
+MONITOR_CHECK_INTERVAL_SECONDS = 0.05
+
+def init_mraa_button(pin):
+    """
+    Instantiate a new mraa.Gpio instance on `pin`.
+    Pull the pin high before returning.
+    Defining this method outside MraaButton
+    so that it is easier to mock from tests.
+    """
+    button = mraa.Gpio(pin)
+    button.dir(mraa.DIR_IN)
+    # pull up before usage
+    mraa_gpio_write(pin, 1)
+    return button
+
+class MraaButton(threading.Thread):
+    """
+    Unlike MraaLED, the button could not be easily monkey patched.
+    Instead, basic button logic is implemented in its entirety.
+
+    MraaButton will trigger #when_held if a button is pressed for
+    hold_seconds. #when_held will only fire once if the button
+    remains held down. If a button is pressed down for shorter,
+    than hold_seconds, #when_held is not called.
+    """
+    
+    def __init__(self, pin, hold_seconds):
+        """
+        pin refers to the pin number, not GPIO number:
+        https://wiki.radxa.com/Rockpi4/hardware/gpio#GPIO_number
+        """
+        super(MraaButton, self).__init__()
+        self.daemon = True
+        self.cancelled = False
+        self.button = init_mraa_button(pin)
+        self.hold_seconds = hold_seconds
+        self.when_held = None
+        self.reset_pressed_state()
+
+    def run(self):
+        while not self.cancelled:
+            if not self.is_pressed():
+                self.reset_pressed_state()
+            else:
+                self.process_click()
+
+            time.sleep(MONITOR_CHECK_INTERVAL_SECONDS)
+
+    def process_click(self):
+        """
+        If this is the first time this press has been detected,
+        record the time for future reference. If the press has
+        already been detected, attempt to trigger #when_held.
+        """
+        
+        if self.last_pressed_at == None:
+            LOGGER.debug("Button pressed for first time")
+            self.last_pressed_at = time.time()
+        else:
+            self.trigger_when_held_after_hold_seconds()
+
+    def trigger_when_held_after_hold_seconds(self):
+        """
+        If #when_held has not been invoked due to this press,
+        and it is defined, invoke it.
+        """
+        
+        if self.when_held and not self.press_already_registered:
+            elapsed = time.time() - self.last_pressed_at
```

Comment:
Should refactor into a helper method.

Reply:
Done: https://github.com/NebraLtd/hm-config/pull/158/files#diff-8026bd004ed0acdc553bd2091db22d9e8a894a353fa081f911964555c45185a9R89